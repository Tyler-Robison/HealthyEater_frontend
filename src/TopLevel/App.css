.App {
  text-align: center;
}

.delete-btn {
  background-color: red;
  color: white;
  border-radius: 20%;
  padding-top: 0px;
  padding: 0px 0.3rem;
  border: 1px black solid;
}

.general-btn {
  padding: 4px 14px;
  border-radius: 6px;
  border: none;
  color: #fff;
  background: linear-gradient(180deg, #4B91F7 0%, #367AF6 100%);
}

.general-btn-green {
  padding: 4px 14px;
  border-radius: 6px;
  border: none;
  color: #fff;
  background: linear-gradient(180deg, #3cc068 0%, #177247 100%);
}

.general-btn-red {
  padding: 4px 14px;
  border-radius: 6px;
  border: none;
  color: #fff;
  background-color: red;
}

button:active {
  transform: scale(1.2);
}

html {
  height: 100vh;
}

body {
  height: 100%;
}

#root {
  height: 100%;
}

.App {
  height: 100%;
}

.error-msg {
  color: red;
}








