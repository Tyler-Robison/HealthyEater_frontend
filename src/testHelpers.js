const mealplanArr = [
    { id: 1, day: 'Mon', recipe_id: 1111, name: 'bacon', ww_points: 20 },
    { id: 2, day: 'Mon', recipe_id: 2222, name: 'eggs', ww_points: 10 },
    { id: 3, day: 'Tues', recipe_id: 3333, name: 'salad', ww_points: 5 }
];

const recipeArr = [
    { name: 'bacon', recipe_id: 1111, ww_points: 20 },
    { name: 'eggs', recipe_id: 2222, ww_points: 10 },
    { name: 'salad', recipe_id: 3333, ww_points: 5 },
    { name: 'bread', recipe_id: 4444, ww_points: 7 },
];

// mock currentUser
const testUser = {
    id: 1,
    username: "Tyler",
    isAdmin: false,
    points: null,
    mealplan: mealplanArr,
    recipes: recipeArr
};

// mock recipe search results
const testRecipes = [
    {
        id: 1234,
        title: 'Scrambled Eggs',
        usedIngredients: [{ name: 'eggs' }, { name: 'otherIngredient' }],
        image: 'https://spoonacular.com/recipeImages/716426-312x231.jpg'
    },
    {
        id: 5678,
        title: 'BLT',
        usedIngredients: [{ name: 'bacon' }, { name: 'lettuce' }, { name: 'tomato' }],
        image: 'https://spoonacular.com/recipeImages/716426-312x231.jpg'
    }
];

export {
    testUser,
    testRecipes
}