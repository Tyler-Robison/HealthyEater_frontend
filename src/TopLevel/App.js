import React, { useState, useEffect } from "react";
import './App.css';
// App.css contains a few universal styles that are used throughout app
import UserAPI from "../APIs/userAPI";
import RouteList from './RouteList';
import NavBar from '../Navigation/NavBar';
import jwt from 'jsonwebtoken'
import useLocalStorage from "../customHooks/useLocalStorage";
import GlobalContext from "../context/GlobalContext";
import { useNavigate } from 'react-router-dom';

/** App is the top-level component 
 * 
 * All state relating to user auth is declared here and passed down to child components
 * 
 * sets up Global context to pass state to other components
 * 
 * Renders NavBar and RouteList which collectively contain all other components of the app*/
function App() {
  const [token, setToken] = useLocalStorage('token')
  const [currentUser, setCurrentUser] = useState(null);
  const [recipes, setRecipes] = useState([])

  const logout = () => setToken(null);
  const login = token => setToken(token);
  const navigate = useNavigate();
  /** logs user into pre-made guest account
   * 
   * username: 'guest'
   * 
   * password: 'password' */
  const guestLogin = async () => {
    const res = await UserAPI.login({ password: 'password', username: 'guest' })
    login(res.token);
    navigate('/')
  }

  useEffect(() => {
    const loginLogout = async () => {
      if (token && token.length !== 0) {

        // token contains id, iat (issued at)
        const id = jwt.decode(token).id

        // userInfo contains:
        // Auth information (id)
        // user's calculated points
        // user's saved recipes
        // user's mealplan
        const userInfo = await UserAPI.getUserInfo(id, token)
        setCurrentUser(userInfo);
      }
      else {
        setCurrentUser(null)
      }
    }
    loginLogout()
  }, [token])


  const providerObj = {
    currentUser,
    setCurrentUser,
    token,
    login,
    recipes,
    setRecipes
  }


  return (
    <div className="App">
      <GlobalContext.Provider value={providerObj}>
        <div className="main-container d-flex h-100 flex-column">
          <NavBar logout={logout} guestLogin={guestLogin} />
          <RouteList />
        </div>
      </GlobalContext.Provider>
    </div>
  );
}

export default App;
