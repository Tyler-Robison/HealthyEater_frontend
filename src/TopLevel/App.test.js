import { render, screen, fireEvent, waitFor, getByText } from '@testing-library/react';

import App from './App'
import { MemoryRouter } from "react-router-dom";
import { testUser, testRecipes } from '../testHelpers'
import mockedAxios from 'axios';

const mockUserData = {
  data: { user: testUser }
}

const mockTokenData = {
  data:
    { token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6Im5ld1VzZXIiLCJpc0FkbWluIjpmYWxzZSwiaWQiOjEyLCJpYXQiOjE2NDY4NjIzMDF9.JJqdcVY4iUqVA3dkNTvSTImGQQbetu0p8pQ7cXHgSd0' }
}

it("renders without crashing", function () {
  <MemoryRouter>
    <App />
  </MemoryRouter>
});

it("matches snapshot", function () {

  const container = render(
    <MemoryRouter initialEntries={['/']}>
      <App />
    </MemoryRouter>
  );
  expect(container.asFragment()).toMatchSnapshot();
});


test("can register new user, starting from homepage", async function () {
  // mock of the token value returned by UserApi.register
  mockedAxios.post.mockResolvedValueOnce(mockTokenData);

  // mock of currentUser by returned by UserAPI.getUserInfo
  mockedAxios.get.mockResolvedValueOnce(mockUserData);

  render(
    <MemoryRouter initialEntries={['/']}>
      <App />
    </MemoryRouter>
  );

  // proves we're on homepage
  expect(screen.getByText('Welcome to Healthy-Eater!')).toBeInTheDocument();

  const signupAnchor = screen.getByText('Signup')
  fireEvent.click(signupAnchor)

  // moved from homepage to sign-in page
  expect(screen.queryByText('Welcome to Healthy-Eater!')).not.toBeInTheDocument();
  expect(screen.getByText('Enter Username/Password')).toBeInTheDocument();

  const signupBtn = screen.getByText('Sign-up');
  const usernameInput = screen.getByLabelText('Username');
  const passwordInput = screen.getByLabelText('Password');

  fireEvent.change(usernameInput, { target: { value: 'Tyler' } });
  fireEvent.change(passwordInput, { target: { value: 'password' } });
  fireEvent.click(signupBtn);

  // navigated back to homepage, showing user-specific welcome back msg
  await waitFor(() => {
    expect(mockedAxios.post.mockResolvedValueOnce(mockTokenData)).toHaveBeenCalledTimes(1)
    expect(mockedAxios.get.mockResolvedValueOnce(mockUserData)).toHaveBeenCalledTimes(1)
    expect(screen.getByText('Welcome back Tyler!')).toBeInTheDocument();
  });
});

test("Sign-up fails with invalid password, correct message displays", async function () {
  // mock of the token value returned by UserAPI.register
  mockedAxios.post.mockResolvedValueOnce(mockTokenData);

  // mock of currentUser by returned by UserAPI.getUserInfo
  mockedAxios.get.mockResolvedValueOnce(mockUserData);

  render(
    <MemoryRouter initialEntries={['/']}>
      <App />
    </MemoryRouter>
  );

  const signupAnchor = screen.getByText('Signup')
  fireEvent.click(signupAnchor)

  const signupBtn = screen.getByText('Sign-up');
  const usernameInput = screen.getByLabelText('Username');
  const passwordInput = screen.getByLabelText('Password');

  fireEvent.change(usernameInput, { target: { value: 'Tyler' } });
  fireEvent.change(passwordInput, { target: { value: 'pwd' } });
  fireEvent.click(signupBtn);

  // registration failure, correct error msg displayed
  await waitFor(() => {
    expect(screen.queryByText('Welcome back Tyler!')).not.toBeInTheDocument();
    expect(screen.getByText('Password must be between 5 to 30 characters')).toBeInTheDocument();
  });
});


const registerNewUser = async (password = 'password') => {
  mockedAxios.post.mockResolvedValueOnce(mockTokenData);
  mockedAxios.get.mockResolvedValueOnce(mockUserData);

  render(
    <MemoryRouter initialEntries={['/']}>
      <App />
    </MemoryRouter>
  );

  const signupAnchor = screen.getByText('Signup')
  fireEvent.click(signupAnchor)

  const signupBtn = screen.getByText('Sign-up');
  const usernameInput = screen.getByLabelText('Username');
  const passwordInput = screen.getByLabelText('Password');

  fireEvent.change(usernameInput, { target: { value: 'Tyler' } });
  fireEvent.change(passwordInput, { target: { value: password } });
  fireEvent.click(signupBtn);
}


test("new user can sign-up then access site functionality", async function () {
  registerNewUser()

  // Don't proceed until registration complete. 
  await waitFor(() => {
    expect(screen.getByText('Welcome back Tyler!')).toBeInTheDocument();
  });

  const recipeAnchor = screen.getAllByText('Find Recipes')[0]
  expect(recipeAnchor).toBeInTheDocument
  fireEvent.click(recipeAnchor)

  const addIngredientBtn = screen.getByText('Add Ingredient')
  const ingredientInput = screen.getByPlaceholderText('Enter Ingredient')
  expect(addIngredientBtn).toBeInTheDocument()
  expect(ingredientInput).toBeInTheDocument()

  fireEvent.change(ingredientInput, { target: { value: 'Cheese' } });
  fireEvent.click(addIngredientBtn)

  // only signed-in users can access this page and add ingredients
  await waitFor(() => {
    expect(screen.getByText('Cheese')).toBeInTheDocument();
  });

  fireEvent.change(ingredientInput, { target: { value: 'Olives' } });
  fireEvent.click(addIngredientBtn)

  await waitFor(() => {
    expect(screen.getByText('Cheese')).toBeInTheDocument();
    expect(screen.getByText('Olives')).toBeInTheDocument();
  });
});

test("user denied site access on failed sign-up", async function () {
  const mockedUsedNavigate = jest.fn();

  jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
  }));

  // password must be 5 characters minimum
  registerNewUser('bad')

  // registerNewUser failed because bad password
  await waitFor(() => {
    expect(screen.getByText('Password must be between 5 to 30 characters')).toBeInTheDocument();
  });

  // Because registration failed, find recipes link isn't showing on screen
  // We can still attempt to manually navigate to that URL
  // User will be re-directed back to homepage. 
  mockedUsedNavigate('/find_recipes')

  // Didn't get to GetRecipes component because user isn't logged in
  expect(screen.queryByText('Ingredients')).not.toBeInTheDocument()
});


