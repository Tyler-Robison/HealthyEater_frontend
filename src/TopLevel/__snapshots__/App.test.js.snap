// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="main-container d-flex h-100 flex-column"
    >
      <nav
        class="NavBar navbar navbar-expand-lg navbar-light"
      >
        <span
          class="NavBar-Brand navbar-brand"
        >
          <a
            aria-current="page"
            class="nav-link active"
            href="/healthyeater"
          >
            Healthy-Eater
            <img
              class="brand-img"
              height="28px"
              src="lettuce.png"
              width="28px"
            />
          </a>
        </span>
        <div
          class="NavBar-Container container"
        >
          <a
            class="nav-link signup "
            href="/signup"
          >
            Signup
          </a>
          <button
            class="general-btn mt-2"
          >
            Continue as Guest
          </button>
          <a
            class="nav-link login"
            href="/login"
          >
            Login
          </a>
        </div>
      </nav>
      <div
        class="Home"
      >
        <div
          class="row mt-3"
        >
          <div
            class="Home-div col-12"
          >
            <h1
              class="Home-h1"
            >
              Welcome to Healthy-Eater!
            </h1>
            <p
              class="Home-para"
            >
               You aren't logged in, 
              <a
                href="/login"
              >
                <b>
                  Login
                </b>
              </a>
               or 
              <a
                href="/signup"
              >
                <b>
                  Sign-up
                </b>
              </a>
               to access site
            </p>
            <p
              class="Home-para"
            >
              <b>
                Find Recipes:
              </b>
               Search for recipes based on ingredients and nutrients
            </p>
            <p
              class="Home-para"
            >
              <b>
                Saved Recipes:
              </b>
               View or delete your saved recipes
            </p>
            <p
              class="Home-para"
            >
              <b>
                Mealplan:
              </b>
               Create your mealplan using saved recipes
            </p>
            <p
              class="Home-para"
            >
              <b>
                Calculate Points:
              </b>
               Determine your weekly allotment of Weight Watcher Smart Points
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
